url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/common/SwingComImageGenerator.java

```diff
@@ -0,0 +1,190 @@
+				logger.debug("generateJPEGImage(Dimension, JComponent[], Rectangle[], String) - Start to paint component " + curCom.getClass().getName() + " to image:" + bounds); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			curCom.paint(imageGraphic);
+			if(bounds.x > 0 || bounds.y > 0){
+				// Copy may be not effective.
+				imageGraphic.copyArea(0, 0, bounds.width, bounds.height, bounds.x, bounds.y);
+			}
+		}
+		try{
+			File imgFile = new File(filePath);
+			FileOutputStream fos = new FileOutputStream(imgFile);
+			BufferedOutputStream bos = new BufferedOutputStream(fos);
+			JPEGImageEncoder encoder = JPEGCodec.createJPEGEncoder(bos);
+			encoder.encode(image);
+			bos.close();
+			if (logger.isDebugEnabled()) {
+				logger.debug("generateJPEGImage(Dimension, JComponent[], Rectangle[], String) - Image:" + filePath + " generated"); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+		}catch (Exception ex) {
+			logger.error("generateJPEGImage(Dimension, JComponent[], Rectangle[], String)", ex); //$NON-NLS-1$
+			throw new IOException(ex.getMessage() + ex.getStackTrace());
+		}
+	}
+	
+	
+	/**
+	 * 
+	 * @param component
+	 * @param filePath
+	 * @throws Exception
+	 */
+	public void genPNGImage(JComponent component, String filePath) throws IOException{
+		Rectangle bounds = component.getBounds();
+		this.genPNGImage(new Dimension(bounds.width, bounds.height), new JComponent[]{component}, new Rectangle[]{bounds}, filePath);
+	}
+
+	/**
+	 * 
+	 * @param size
+	 * @param components
+	 * @param areas
+	 * @param filePath
+	 * @throws IOException
+	 */
+	public void genPNGImage(Dimension size, JComponent[] components, Rectangle[] areas, String filePath) throws IOException{
+		logger.debug("Generate JPEG image file");
+		if(components.length != areas.length){
+			throw new RuntimeException("The specified components is not fulfil the requirement");
+		}
+		if(filePath == null){
+			String dir = System.getProperty("user.dir") + System.getProperty("file.separator") + "image";
+			filePath = dir + System.getProperty("file.separator") + "ganttchart.jpg";
+		}
+		checkAndCreateDirForFilePath(filePath);
+		BufferedImage image = new BufferedImage(size.width, size.height, BufferedImage.TYPE_INT_RGB);
+		Graphics imageGraphic = image.getGraphics();
+		// Background is white.
+		imageGraphic.setColor(Color.WHITE);
+		imageGraphic.fillRect(0, 0, size.width, size.height);
+
+		for (int i = 0; i < components.length; i++) {
+			JComponent curCom = components[i];
+			Rectangle bounds = areas[i];
+			if (logger.isDebugEnabled()) {
+				logger.debug("Start to paint component " + curCom.getClass().getName() + " to image:" + bounds); //$NON-NLS-1$ //$NON-NLS-2$
+			}
+			curCom.paint(imageGraphic);
+			if(bounds.x > 0 || bounds.y > 0){
+				// Copy may be not effective.
+				imageGraphic.copyArea(0, 0, bounds.width, bounds.height, bounds.x, bounds.y);
+			}
+		}
+		try{
+			File imgFile = new File(filePath);
+			ImageOutputStream ios = new FileImageOutputStream(imgFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *ios*. The resource is referenced at line 162. The resource is closed at line 167. There are other execution paths that do not contain closure statements, for example, when ImageWriter.write throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).